README.md
# OOD

[![CodeQL](https://github.com/JasonMahdjoub/OOD/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/JasonMahdjoub/OOD/actions/workflows/codeql-analysis.yml)

OOD is a decentralized and asynchrone ORM (Object-Relational Mapping):
* Compatible with Java 8 and newer
* drivers for MySQL, PostgreSQL, H2, H2 for Android (>=Android 8, >=API26), HSQLDB. When using Android driver of H2, please initialize android context by affecting variable 'com.distrimind.util.harddrive.AndroidHardDriveDetect.context'
* Transparent decentralized and secured database synchronization
  * Automatically synchronize database with other authorized peers, even if peers where disconnected (asynchronous mode).
  * Manage eventual conflicts during database synchronization
  * Automatically backup database and it historical to authorized servers and use end-to-end encryption : the servers are not able to read database in clear. 
  * Autotatically synchronize peers through server side database backup if peers are not connected (asynchronous mode)
  * Permit to revert database to an old version with a given time, and synchronize the restoration toward others peers and servers. It is possible to choose data source from local backup, other peers backup, or server backup.
  * Backups cannot be removed into servers from peers for a certain period. They cannot also be removed from a peer into a distant peer. This permit to garuantee the restoration even after an attack of a ransomware. 
  * Permit to restore database of peers thanks to other peers or thanks to servers
  * Automatically synchronize peers though intermediate peers when direct connections are not possible
  * Use encryption profile provider and to manage different version of secret keys in the case where a secret key was compromised
* pseudo SQL querries possible (where clause)



# Changes

[See historical of changes](./changelog.md)

# How to use it ?
## With Gradle :

Adapt into your build.gradle file, the next code :
```
	...
	dependencies {
		...
		api(group:'com.distrimind.ood', name: 'OOD', version: '3.3.0-STABLE')
		//choose one of these optional drivers for H2 database
		api(group:'com.h2database', name: 'h2', version: '2.1.210')
		//choose this optional driver for HSQLdb
		api(group:'org.hsqldb', name: 'hsqldb', version: '2.6.1')
		//choose this optional driver for MySQL
		api(group: 'mysql', name: 'mysql-connector-java', version: '8.0.28')
		//choose this optional driver for PostgreSQL
		api(group: 'org.postgresql', name: 'postgresql', version: '42.3.2')
	}
	...
```
Librairies are available on Maven Central. You can check signatures of dependencies with this [public GPG key](./gpg_key.asc). You can also use the next repository : 
```
	...
	repositories {
		...
		maven {
	       		url "https://artifactory.distri-mind.fr:443/artifactory/gradle-release/"
	   	}
		...
	}
	...
```
To know what last version has been uploaded, please refer to versions availables into [this repository](https://artifactory.distri-mind.fr/artifactory/DistriMind-Public/com/distrimind/ood/OOD/)
## With Maven :
Adapt into your pom.xml file, the next code :
```
	...
	<project>
		...
		<dependencies>
			...
			<dependency>
				<groupId>com.distrimind.ood</groupId>
				<artifactId>OOD</artifactId>
				<version>3.3.0-STABLE</version>
			</dependency>
			<!-- choose one of these optional drivers for H2 database-->
				<dependency>
					<groupId>com.h2database</groupId>
					<artifactId>h2</artifactId>
					<version>2.1.210</version>
				</dependency>

			<!-- choose one of these optional drivers for HSQLDB-->
				<dependency>
					<groupId>org.hsqldb</groupId>
					<artifactId>hsqldb</artifactId>
					<version>2.6.1</version>
				</dependency>
			<!-- choose this optional driver for MySQL-->
				<dependency>
					<groupId>mysql</groupId>
					<artifactId>mysql-connector-java</artifactId>
					<version>8.0.28</version>
				</dependency>
			<!-- choose this optional driver for PostgreSQL-->
				<dependency>
					<groupId>org.postgresql</groupId>
					<artifactId>postgresql</artifactId>
					<version>42.3.2</version>
				</dependency>
			...
		</dependencies>
		...
	</project>
	...
```	
Librairies are available on Maven Central. You can check signatures of dependencies with this [public GPG key](./gpg_key.asc). You can also use the next repository : 
```
	...
	<repositories>
		...
		<repository>
			<id>DistriMind-Public</id>
			<url>https://artifactory.distri-mind.fr:443/artifactory/gradle-release/</url>
		</repository>
		...
	</repositories>
	...		
```
To know what last version has been uploaded, please refer to versions availables into [this repository](https://artifactory.distri-mind.fr/artifactory/DistriMind-Public/com/distrimind/ood/OOD/)

###### Requirements under Ubuntu/Debian :
  * Please install the package ethtool, rng-tools, mtr(only debian)


changelog.md
Object Oriented Database
========================
3.3.0 STABLE (Build: 6141) (from 01/04/2013 to 10/01/2023)

# Creator(s):
* Jason MAHDJOUB

# Developer(s):
* Jason MAHDJOUB (Entered in the team at 01/04/2013)

# Changes:


### 3.3.0 STABLE (10/01/2023)
#### New feature(s)
* Add serialization of hashed values
* Add possibility to check string regex patterns with table fields. See annotation Field.
* Manage StringBuilder and StringBuffer types.
#### Internal change(s)
* Update Java language level to Java 11
* Update Utils to 5.25.2 STABLE


### 3.2.1 STABLE (23/12/2022)
#### Internal change(s)
* Fix spelling errors
* Update Utils to 5.24.1


### 3.2.0 STABLE (08/12/2022)
#### New feature(s)
* Add database tasks manager, that permit to schedule periodic tasks executed into the database. See package com.distrimind.ood.database.tasks
* Add possibility to specify a context with DatabaseWrapper. See DatabaseFactory.setContext(Object)
* Add operator IN/NOT IN into pseudo SQL interpreter, that permit to test an identifier with a collection
* Implements mathematical operator (+, -, *, /, %) with pseudo SQL queries. Theses operators work with Numbers, and Date values.
#### Internal change(s)
* Target java compatibility is set to Java 11 but source code still use Java 8
* Update Utils to 5.24.0
* Update dependencies
* Significant optimizations of pseudo SQL interpreter.


### 3.1.26 STABLE (07/04/2022)
#### Internal change(s)
* Update Utils to 5.23.5 STABLE


### 3.1.22 STABLE (05/04/2022)
#### Internal change(s)
* Update Utils to 5.23.2 STABLE
* Add lacking files into generated libraries.


### 3.1.21 STABLE (05/04/2022)
#### Internal change(s)
* Update Utils to 5.23.1 STABLE
* Remove finalize methods into all classes, and these classes using Cleanable API


### 3.1.20 STABLE (29/03/2022)
#### Internal change(s)
* Update Utils to 5.22.4 STABLE


### 3.1.19 STABLE (25/03/2022)
#### Internal change(s)
* Update Utils to 5.22.3 STABLE


### 3.1.17 STABLE (08/02/2022)
#### Internal change(s)
* Update Utils to 5.22.2 STABLE
* Update HSQLDB to 2.6.1
* Update H2 to 2.1.210
* Update mysql-connector-java 8.0.28
* Update postgresql 42.3.2
* Minimum Android version to use H2 is Android 8.0 or API26
* Update URLs
#### Bug fixe(s)
* Check if certificate is given when initialising database configurations that are synchronized with central database backup


### 3.1.16 STABLE (24/01/2022)
#### Internal change(s)
* Update Utils to 5.22.1 STABLE


### 3.1.15 STABLE (24/01/2022)
#### New feature(s)
* Manage WrappedString and WrappedData classes into field accessors and into queries
#### Internal change(s)
* Update Utils to 5.22.0 STABLE


### 3.1.14 STABLE (22/12/2021)
#### Internal change(s)
* Update Utils to 5.21.7 STABLE


### 3.1.13 STABLE (22/12/2021)
#### Internal change(s)
* Update Utils to 5.21.6 STABLE


### 3.1.12 STABLE (16/12/2021)
#### Internal change(s)
* Update Utils to 5.21.5 STABLE


### 3.1.11 STABLE (15/12/2021)
#### Internal change(s)
* Update Gradle to 7.3.1


### 3.1.10 STABLE (09/12/2021)
#### Internal change(s)
* Update Utils to 5.21.3


### 3.1.9 STABLE (03/12/2021)
#### Internal change(s)
* Little corrections


### 3.1.8 STABLE (03/12/2021)
#### Internal change(s)
* Update Utils to 5.21.2 STABLE


### 3.1.7 STABLE (02/11/2021)
#### Internal change(s)
* Update Utils to 5.21.0 STABLE
* Update MySQL JDBC driver to 8.0.27
* Update PostgreSQL JDBC driver to 42.3.1


### 3.1.6 STABLE (18/10/2021)
#### Internal change(s)
* Update Utils to 5.20.6 STABLE


### 3.1.3 STABLE (13/10/2021)
#### Internal change(s)
* Update Utils to 5.20.3 STABLE


### 3.1.2 STABLE (13/10/2021)
#### Internal change(s)
* Update Utils to 5.20.2 STABLE
* Better log formatting
#### Bug fixe(s)
* Fix issue when sending messages from central server to disconnected peers
* Fix issue with some messages that where not sent from one server, when several servers where used


### 3.1.1 STABLE (01/10/2021)
#### Internal change(s)
* Update Utils to 5.19.7 STABLE
#### Bug fixe(s)
* Fix issue and better update last restoration timestamp of a database


### 3.1.0 STABLE (30/09/2021)
#### New feature(s)
* Add file manager
* Permit indirect initial synchronization between peers through central database backup
* Automatically restore database from server during first connection
#### Internal change(s)
* Remove concatenated sql queries
* Clean code
#### Bug fixe(s)
* Fix issue and avoid creating backup reference when the database is empty. Create first backup reference at the start of the transaction, before the first queries
* Fix issue when loading all fields instead of loading only primary keys from backup
* Fix issue when database is restored to an old version, and when old synchronization message from other peers are received just after the restoration. Now these messages are ignored when timestamp are lower than the last restoration time.
* Fix synchronization lock when two peers must apply synchronization through central database backup and whereas initial database is empty into each peer.


### 3.1.0 BETA 1 (07/07/2021)
#### New feature(s)
* Add function void DatabaseFactory.setEncryptionProfileProviders(EncryptionProfileProviderFactory, SecureRandomType randomType, byte[], byte[])
* Add function void DatabaseFactory.setEncryptionProfileProviders(EncryptionProfileProviderFactory, SecureRandomType randomType)
* Add class CentralDatabaseBackupReceiverFactory
* Add function CentralDatabaseBackupReceiver.sendMessageFromThisCentralDatabaseBackup(MessageComingFromCentralDatabaseBackup)
* Add function BackupRestoreManager.hasNonFinalFiles()
* Add function DatabaseWrapper.getNextPossibleEventTimeUTC()
#### Internal change(s)
* Update Utils to 5.18.5
#### Bug fixe(s)
* Make OOD compatible with MadKitLanEdition
* Fix NullPointerException into DatabaseConfiguration class
* Fix bad using of central database backup when it is not used
* Fix issue with notification of local host initialization

OOD/src/com/distrimind/ood/OOD.java

/*
Copyright or © or Copr. Jason Mahdjoub (01/04/2013)

jason.mahdjoub@distri-mind.fr

This software (Object Oriented Database (OOD)) is a computer program 
whose purpose is to manage a local database with the object paradigm 
and the java langage 

This software is governed by the CeCILL-C license under French law and
abiding by the rules of distribution of free software.  You can  use, 
modify and/ or redistribute the software under the terms of the CeCILL-C
license as circulated by CEA, CNRS and INRIA at the following URL
"http://www.cecill.info". 

As a counterpart to the access to the source code and  rights to copy,
modify and redistribute granted by the license, users are provided only
with a limited warranty  and the software's author,  the holder of the
economic rights,  and the successive licensors  have only  limited
liability. 

In this respect, the user's attention is drawn to the risks associated
with loading,  using,  modifying and/or developing or reproducing the
software by the user in light of its specific status of free software,
that may mean  that it is complicated to manipulate,  and  that  also
therefore means  that it is reserved for developers  and  experienced
professionals having in-depth computer knowledge. Users are therefore
encouraged to load and test the software's suitability as regards their
requirements in conditions enabling the security of their systems and/or 
data to be ensured and,  more generally, to use and operate it in the 
same conditions as regards security. 

The fact that you are presently reading this means that you have had
knowledge of the CeCILL-C license and that you accept its terms.
 */
package com.distrimind.ood;

import java.io.InputStream;
import java.util.Calendar;

import com.distrimind.util.export.License;
import com.distrimind.util.version.Description;
import com.distrimind.util.version.Person;
import com.distrimind.util.version.PersonDeveloper;
import com.distrimind.util.version.Version;

/**
 * 
 * @author Jason Mahdjoub
 * @version 2.0.0
 * @since OOD 1.0
 */
public class OOD
{
    public static final Version VERSION;
    public static final License LICENSE=new License(License.PredefinedLicense.CeCILL_C_v1_0);
    static
    {
	Calendar c1=Calendar.getInstance();
	c1.set(2013, 3, 1);
	Calendar c2=Calendar.getInstance();
	c2.set(2017, 6, 27);
	VERSION=new Version("Object Oriented Database", "OOD", 2, 0, 0, Version.Type.Beta, 8, c1.getTime(), c2.getTime());
	try
	{
	    InputStream is=OOD.class.getResourceAsStream("build.txt");
	    VERSION.loadBuildNumber(is);
	    VERSION.addCreator(new Person("mahdjoub", "jason"));
	    Calendar c=Calendar.getInstance();
	    c.set(2013, 3, 1);
	    VERSION.addDeveloper(new PersonDeveloper("mahdjoub", "jason", c.getTime()));
	
	    c=Calendar.getInstance();
	    c.set(2017, 6, 27);
	    Description d=new Description(2,0,0,Version.Type.Beta, 8, c.getTime());
	    d.addItem("All decentralized database synchronization tests are OK with HSQLDB (but not with Derby DB).");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2017, 6, 5);
	    d=new Description(2,0,0,Version.Type.Beta, 7, c.getTime());
	    d.addItem("Tests synchro between two direct peers OK.");
	    d.addItem("Correcting a problem of transaction lock.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2017, 6, 2);
	    d=new Description(2,0,0,Version.Type.Beta, 6, c.getTime());
	    d.addItem("Adding not null possibility for each sub field.");
	    d.addItem("Begin debug decentralized database.");
	    d.addItem("Optimizing tables junction.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2017, 5, 23);
	    d=new Description(2,0,0,Version.Type.Beta, 5, c.getTime());
	    d.addItem("Adding not null possibility for each sub field.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2017, 5, 14);
	    d=new Description(2,0,0,Version.Type.Beta, 4, c.getTime());
	    d.addItem("Debugging SQL interpreter.");
	    d.addItem("Adding not null possibility for each sub field.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2017, 5, 1);
	    d=new Description(2,0,0,Version.Type.Beta, 3, c.getTime());
	    d.addItem("Adding database factory.");
	    d.addItem("Correcting a bug in database connections/deconnections.");
	    d.addItem("Updating Utils to 2.8.0.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2017, 4, 26);
	    d=new Description(2,0,0,Version.Type.Beta, 2, c.getTime());
	    d.addItem("Debuging transaction's use.");
	    d.addItem("Adding pagined queries.");
	    d.addItem("Adding ordered queries.");
	    d.addItem("Adding records numbers queries.");
	    d.addItem("Correcting a bug with multi fields match querries.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2017, 4, 24);
	    d=new Description(2,0,0,Version.Type.Beta, 1, c.getTime());
	    d.addItem("Adding database synchronisation possibility between different peers (unstable).");
	    d.addItem("Cleaning DatabaseWrapper functions.");
	    d.addItem("Correcting a lock problem into DatabaseWrapper.runSynchronizedTransaction(SynchronizedTransaction) function.");
	    d.addItem("Optimizing Decentralized IDs.");
	    d.addItem("Adding SQL where treatments (unstable).");
	    d.addItem("Add indexation field possibility (see annotation Field).");
	    d.addItem("Updating to Utils 2.7.1.");
	    d.addItem("Downgrading to Derby 10.11.1.1 (JDK 7 Compatible).");
	    d.addItem("Adding DatabaseConfiguration class.");
	    d.addItem("Adding DatabaseCreationCallable class.");
	    d.addItem("Changing the database loading policy : enabling transfer data from old database to new database.");
	    d.addItem("Minimizing code duplication of tests.");
	    d.addItem("Adding remove database possibility.");
	    d.addItem("Use of savepoints.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2017, 2, 7);
	    d=new Description(1,9,7,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 2.5.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2017, 2, 4);
	    d=new Description(1,9,6,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 2.4.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2017, 1, 7);
	    d=new Description(1,9,5,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 2.3.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2017, 0, 5);
	    d=new Description(1,9,4,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 2.2.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 11, 31);
	    d=new Description(1,9,3,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 2.1.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 11, 26);
	    d=new Description(1,9,2,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Correcting a problem of data integrity check.");
	    d.addItem("Updating Derby DB to 10.13.1.1.");
	    d.addItem("Solving a problem of multithread execution into windows (SQLNonTransientConnectionException).");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 11, 24);
	    d=new Description(1,9,1,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to version 2.0.1.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 11, 19);
	    d=new Description(1,9,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to version 2.0.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 11, 6);
	    d=new Description(1,8,1,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to version 1.9.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 10, 30);
	    d=new Description(1,8,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Using ACID transactions instead of semaphores.");
	    d.addItem("Adding HSQLDBConcurrencyControl class.");
	    d.addItem("Adding memory refresh interval (see LoadToMemory annotation).");
	    d.addItem("Adding the possibility to set composed fields by adding @Field to a class.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 9, 13);
	    d=new Description(1,7,1,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 1.8.0.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 8, 19);
	    d=new Description(1,7,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Correcting bug into FieldAccessor (null pointer exception).");
	    d.addItem("Adding enum support.");
	    d.addItem("Adding Decentralized ID support.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 7, 29);
	    d=new Description(1,6,4,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 1.7.2.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 7, 23);
	    d=new Description(1,6,3,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 1.7.1.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 6, 4);
	    d=new Description(1,6,2,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating Utils to 1.7.");
	    d.addItem("Updating to Common-Net 3.5.");
	    d.addItem("Updating to HSDQLDB 3.3.4.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 5, 10);
	    d=new Description(1,6,1,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Correction a bug into the constructor of ByteTabConvertibleFieldAccessor.");
	    d.addItem("Adding version tests.");
	    d.addItem("Changing license to CECILL-C.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 2, 11);
	    d=new Description(1,6,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating with Utils 1.6.");
	    d.addItem("Adding database backup tools.");
	    d.addItem("renaming alterRecord functions to updateRecord.");
	    d.addItem("Adding functions Table.addRecord(record), Table.updateRecord(record).");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 2, 4);
	    d=new Description(1,5,2,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating with Utils 1.5");
	    d.addItem("Adding encryption keys encoding/decoding.");
	    d.addItem("Correcting bugs with ByteTabConvertibleFieldAccessor class.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 2, 1);
	    d=new Description(1,5,1,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Updating with Utils 1.4.");
	    d.addItem("Adding AllTestsNG.xml file.");
	    VERSION.addDescription(d);
	    
	    c=Calendar.getInstance();
	    c.set(2016, 1, 15);
	    d=new Description(1,5,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding iterator functionality in class Table.");
	    d.addItem("Adding ByteTabObjectConverter class.");
	    d.addItem("Adding DefaultByteTabObjectConverter class.");
	    d.addItem("Adding ByteTabConvertibleFieldAccessor class.");
	    d.addItem("Adding function addByteTabObjectConverter in DatabaseWrapper class.");
	    d.addItem("Adding possibility to use Object tabs as an alternative of use of maps when reffering to fields.");
	    d.addItem("Optimizing use of SQL database.");
	    d.addItem("Linking with Utils 1.3.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2016, 1, 14);
	    d=new Description(1,4,1,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding some close statements corrections.");
	    d.addItem("Adding some multi-thread optimisations.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2016, 1, 8);
	    d=new Description(1,4,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("One databse is associated to one package. Now, its is possible to load several database/packages into the same file.");
	    d.addItem("OOD works now with HSQLDB or Apache Derby.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2016, 1, 5);
	    d=new Description(1,3,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Adding dependency with Utils and updating OOD consequently.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2016, 1, 1);
	    d=new Description(1,2,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Correcting some bugs into the documentation.");
	    d.addItem("Upgrading to HSQLDB 2.3.3 and Commons-Net 3.4.");
	    VERSION.addDescription(d);

	    c=Calendar.getInstance();
	    c.set(2013, 10, 18);
	    d=new Description(1,1,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Improving function Table.alterRecords(...) and class AlterRecordFilter (see documentation).");
	    VERSION.addDescription(d);
	
	    c=Calendar.getInstance();
	    c.set(2013, 3, 24);
	    d=new Description(1,0,0,Version.Type.Stable, 0, c.getTime());
	    d.addItem("Releasing Oriented Object Database as a stable version.");
	    VERSION.addDescription(d);
	}
	catch(Exception e)
	{
	    e.printStackTrace();
	}
    }    
    
}
